### DISCUSSION TITLE
Checkbox forwarded Ref should set on Input field and not Button

### DISCUSSION THREAD

**Discussion by emadabdulrahim** [radix-ui/primitives#874] [Help] [answered]
I tried using Checkbox and pass it a ref, but it seems like the ref is being set on the Button element and not the actual input element itself. This means any library expecting to access the native Input event will not work properly.

I believe the Ref should be set on the hidden input element itself.

https://github.com/radix-ui/primitives/blob/c553b2fe6d29f7415f38c13e6bcbfad052e9c8c1/packages/react/checkbox/src/Checkbox.tsx#L76

**Comment by jjenzz** [comment, marked as answer]
This is intentional because the `Checkbox` is a `button` so you will get a `button`. The `input` is only rendered when it is used within a `form` in order to propagate events. If it wasn't for that necessity, we wouldn't render one at all.

Would you mind helping us understand why you need a reference to an `input`, what is your use-case?

**Reply by emadabdulrahim** [reply]
Hi jjenzz, 

I'm using react-hook-form which relies on getting access to the form elements through a ref. When using Radix Checkbox, it fails to read the check event properly, the result is the form doesn't change the value of that checkbox.

**Reply by jjenzz** [reply]
@emadabdulrahim You can control the Radix Checkbox with react-hook-form as follows https://codesandbox.io/s/weathered-worker-huynj?file=/src/App.js

**Reply by emadabdulrahim** [reply]
@jjenzz Thanks for providing a code example. I saw it in their docs but not sure why I didn't give it a try. Cheers!

**Reply by AndresdoSantos** [reply]
@jjenzz Wow, saved!

**Reply by mstykow** [reply]
If anybody is wondering: a similar solution works for working with [Radix' radio group](https://www.radix-ui.com/docs/primitives/components/radio-group).

**Reply by Sanarielsen** [reply]
@jjenzz I really appreciated about your solution! 

I have another doubt, what is correct way to typing "props" in line 16?

Again, thank you very much for your help!!!

**Reply by jjenzz** [reply]
@Sanarielsen i haven't tried it here myself but something like `React.ComponentProps<typeof Controller>` might work

**Reply by Sanarielsen** [reply]
That's right! Its working in the code. Appreciate that! @jjenzz

**Reply by Sanarielsen** [reply]
I'm checked the code and i have a problem with this format.

![image](https://user-images.githubusercontent.com/67290994/234022631-b41185e0-e21a-4e98-8fa1-221feab219c7.png)

Witch format i put in this case to resolve this warning?

**Reply by ozzyfromspace** [reply]
@jjenzz thanks for the codesandbox demo! it works and all, but every time the checkbox is clicked, it causes react to re-render, partially defeating the use of react-hook-form. If you switch to a native input of type checkbox, you can spread the rhf register function directly into the input's props, and everything just works. This is, perhaps, a good reason to expose the input ref that shadows the button. I'm working on a big form with hundreds of inputs and re-rendering on every click just feels bad here. Any advice on how to proceed? I know I'm awfully late to this discussion. Thanks.

**Reply by jjenzz** [reply]
@ozzyfromspace it will re-render the checkbox because the checkbox state has changed which it needs to do otherwise the `<Checkbox.Indicator />` cannot mount/unmount. it does not re-render the whole form (see my codesandbox which re-renders the checkbox only) so if that's what you meant, it might be helpful to share your code to help debug the issue with your particular form.

**Reply by ozzyfromspace** [reply]
@jjenzz hope you're well ☮️. just wanted to mention that a standard input will not re-render with react-hook-form.

I think it may be worth redoing the radix checkbox implementation, because every browser today supports "appearance: none" so inputs of type checkbox are pretty easy to work with now, and users that rely on more checkbox events would be better served.

Is this worth a refactor PR? I'm willing to redo the checkbox with a standard input **without changing the current API surface** if that sounds reasonable. If not that's understandable.

**Reply by cereallarceny** [reply]
@ozzyfromspace I'd love to see this too. I was shocked to see Radix handling Checkboxes this way, considering every other UI framework is largely plug-and-play with common form libraries like react-hook-form. If you have any suggestions for how to fix this outside of a PR to Radix, I'd love to know what you've come up with.

**Reply by ozzyfromspace** [reply]
@cereallarceny I'll create a codesandbox to demo how I'm working with checkboxes. Hopefully you find it useful.

**Reply by ozzyfromspace** [reply]
@cereallarceny here's a simplified version of how I'm working with checkboxes in my current project:

https://codesandbox.io/s/goofy-chaum-q553vp?file=/src/App.tsx

Worth noting, handling things like disabling buttons and such may cause re-renders if you rely on formState (nothing wrong with that, just a thing to be aware of). Also, with this approach, react-hook-form "just works". This is what I could imagine radix-ui's checkbox could be like.

**Reply by cereallarceny** [reply]
Thanks so much for sharing @ozzyfromspace , I'll check it out!

**Reply by ozzyfromspace** [reply]
@cereallarceny my pleasure. Best wishes with your code

**Reply by jjenzz** [reply]
@benoitgrelard it'd be really great if Radix wrote a blog post or something on why we used `button` for checkbox. there is good reason (which Radix are aware of internally) but this gets raised a lot.

> Is this worth a refactor PR? I'm willing to redo the checkbox with a standard input without changing the current API surface if that sounds reasonable. If not that's understandable.

@ozzyfromspace i would say no. tmk, refactoring it to match the current API and DX requirements is not possible. one of the main requirements being that consumers should not need to use `peer-checked` or `input + .indicator` selectors etc. to style the indicator. 

a rule of thumb with Radix is that you should only interface with the public API (the component props) because it will not incur breaking changes. anything that the component renders is considered **private**—an implementation detail—which could change at any time so it's best to avoid targetting anything that relies on the component output/DOM structure (e.g. `input + .indicator`) if you want to avoid breaking changes.

**Reply by emadabdulrahim** [reply]
Another challenge with Radix Checkbox is that `onCheckedChange` doesn't pass input `event`. This introduces all sorts of issues. For example, to add multi-select functionality (like in Gmail), you need access to `event` to see if shift key is pressed.

**Reply by enix79** [reply]
> anything that the component renders is considered private—an implementation detail—**which could change at any time**

@jjenzz, I don't think that the input will disappear any time. So why is it an issue, when you expose the input's ref and make it public? How should we integrate it with RHF?

And what is the "good reason" for using a button? I like Radix UI, but there should be an explanation of the whys. Sometimes I don't have to style the components too heavily, but still need them to be accessible. So if this reason is only about the styling thing, but not the accessibility thing, I would like to know that.

**Comment by equinusocio** [comment]
This is still a huge limitation, because you can only fix this when you have direct access to the components. Many libraries, like react-table, expect a real input element with native API. Using those fake inputs leads into type errors, since is not rendered inside a form.

**Reply by emadabdulrahim** [reply]
I love Radix but I must agree that their Checkbox implementation is suboptimal. There should be no need to have a workaround to access the native element's events and attributes.

**Reply by emadabdulrahim** [reply]
Any component implementation that hides or makes it difficult to access the native element's events is wrong, IMO.

**Reply by equinusocio** [reply]
In this case, there is no a native input at all. All the selection controls like radio, button and toggles are just buttons when rendered outside forms. So, whenever native events and callback are required, they fail. No reason at all to use `button` over native inputs since they are fully customizable now.

**Reply by benoitgrelard** [reply]
> All the selection controls like radio, button and toggles are just buttons when rendered outside forms.

That is correct, if it's outside of a form, there isn't any need for any native component anyway because bubbling only makes sense within forms.

> So, whenever native events and callback are required, they fail

We have a consistent API which lets you control the component's value/checked state, so there should be nothing that is impossible to handle. If there is please let us know.

> No reason at all to use button over native inputs since they are fully customizable now.

As far as I know they aren't. You cannot nest anything inside an input, so users would have to rely on unintuitive sibling CSS selectors to style things or add their own wrappers, etc. This doesn't sit well with the guiding principles behind our primitives.

**Reply by swernerx** [reply]
Would it be too hard to offer an `inputRef` prop like MaterialUI does? I guess this would solve most of these issues. See also: https://mui.com/material-ui/api/text-field/#props:

> Pass a ref to the input element.

**Reply by ryanmr** [reply]
I ran into this issue today with react-hook-form and the radix checkbox. It's a seriously unfortunate shortcoming. @swernerx thanks for the link to the mui example. I looked at the more complex checkbox code in radix. Not sure where to go from there.

**Reply by ozzyfromspace** [reply]
@ryanmr I would imagine when radix first tackled the checkbox, it was notoriously hard to style, so they went with a button implementation.

However, since 2017, basically all browsers have supported the css property "appearance" which when set to "none" removes all browser agent styles on the element and makes it easy for you to work with.

With this in mind, I recommend just using a simple `<input type="checkbox" />` and setting your classes appropriately (e.g. in tailwind, class="appearance-none OTHER_CLASSES")

If you use a native input element of type checkbox, and ensure "appearance: none;" is set in your css, everything "just works" with react hook form.

I'm working on a local govt. website and this is our solution. I still love and use radix for other things, just not the checkbox, and that's okay.

Reference: https://developer.mozilla.org/en-US/docs/Web/CSS/appearance

**Reply by ozzyfromspace** [reply]
@swernerx ideally the radix checkbox would just use the html input element under the hood

**Reply by plam-ml** [reply]
Ran into this. I solved this using the hook controller component

```
import { type ComponentProps, type ReactNode, forwardRef } from 'react';
import { Controller, useFormContext } from 'react-hook-form';

import { Checkbox } from './Checkbox';

type CheckboxFieldProps = {
    name: string;
    label: ReactNode;
} & Omit<ComponentProps<'button'>, 'ref'>;

// eslint-disable-next-line react/display-name
export const CheckboxField = forwardRef<HTMLButtonElement, CheckboxFieldProps>(
    ({ label, name, ...props }, forwardedRef) => {
        const { control } = useFormContext();
        return (
            <Controller
                control={control}
                name={name}
                render={({ field: { onChange, name, value } }) => (
                    <Checkbox
                        label={label}
                        onCheckedChange={onChange}
                        name={name}
                        checked={value}
                        ref={forwardedRef}
                        {...props}
                    />
                )}
            />
        );
    },
);
```

**Reply by enix79** [reply]
@plam-ml - how does it solve the problem? If your "Checkbox" uses Radix UI Checkbox, your forwardedRef will be the button ref, not the input ref.

**Comment by ks-himanshub** [comment]
Actually in my case everything is working fine but I am using this WAVE extension (available in chrome web store),

Which say that a an input element must have a label,

But in case of radix when we use checkbox inside a form tag a input element is rendered but we can't set any attribute to that element (input - rendered inside checkbox)

**Reply by ks-himanshub** [reply]
https://wave.webaim.org/

**Reply by nzwnabdulwahid** [reply]
I am having the same issue right now, are you able to solve this @ks-himanshub ?

**Comment by nuxcodes** [comment]
The `input` element is invisible but has a size attribute. This creates a weird focus ring on a `Checkbox` placed inside a `form` element.

<img width="263" alt="image" src="https://user-images.githubusercontent.com/44252050/229904157-f1be4a02-a82e-4779-9162-74e82a290e82.png">

@jjenzz Any ideas? Thanks.

**Reply by jjenzz** [reply]
@benoit i have noticed some issues with the `input` positioning myself separately. maybe it's worth having an issue to look into all this? if i recall, the intention was for the `input` to overlay the fake one **only** so that the native validation popover is aligned correctly.

https://user-images.githubusercontent.com/175330/233645249-484854ce-2555-41b7-b0ac-1baeec9d4550.mp4

**Reply by benoitgrelard** [reply]
Weird…

We wrote this over 2 years ago:

https://github.com/radix-ui/primitives/blob/45069f85acb2251929e250c0db5b76620a4ecf1b/packages/react/checkbox/src/Checkbox.tsx#L112-L115

But it seems it's not needed? 🤔

**Reply by jjenzz** [reply]
it will be positioned correctly when the parent wrapper isn't `display: flex`. the docs have added the flex to position the label. it'll need to be solved some other way i imagine.

**Reply by mglavall** [reply]
I'm having this issue which the `input` is not positioned on top of the `button` and it's even further away.
In this picture I'm hovering the terms and conditions input. It's so far away it even creates a scroll on the page.
<img width="500" alt="image" src="https://github.com/radix-ui/primitives/assets/23035545/7ca1ba6e-7c6c-439c-9a11-9d72bb79c33f">

**Reply by mglavall** [reply]
More on this: Found out that in my case there is a further parent that is a `display: grid` which if I disable the input positions at the side. So this might be another different issue, but related

**Comment by enix79** [comment]
I read the discussion, but I still do not understand how the accepted answer helps here. We also use RHF and we have to pass ref to the input, not button to get it work :(

**Comment by washingtonsoares** [comment]
If you, like me, just want to integrate with RHF, then this might be helpful:

```tsx

<Controller
  control={control}
  name="hasAcceptedTerms"
  render={({ field: { onChange, value } }) => (
    <Checkbox
      onCheckedChange={onChange}
      checked={value}
      id="terms"
    />
  )}
/>

**Reply by 58bits** [reply]
> If you, like me, just want to integrate with RHF, then this might be helpful:
> 
> ```tsx
> <Controller
>   control={control}
>   name="hasAcceptedTerms"
>   render={({ field: { onChange, value } }) => (
>     <Checkbox
>       onCheckedChange={onChange}
>       checked={value}
>       id="terms"
>     />
>   )}
> />
> ```

This is also what we do at the moment.
